0-module-and-module-source.emu

```diff
@@ -420,151 +476,49 @@ location: https://tc39.es/proposal-compartments/
-            <h1>
-              FinishModuleDynamicImport (
-                _moduleRecord_: a Module Record,
-                _promiseCapability_: a PromiseCapability Record,
-                _innerPromise_: unknown,
-              ): ~unused~
-            </h1>
-            <dl class="header">
-              <dt>description</dt>
-              <dd>FinishModuleDynamicImport completes the process of a dynamic import originally started by an <emu-xref href="#sec-import-calls">`import()`</emu-xref> call for a Module Instance, resolving or rejecting the promise returned by that call as appropriate according to _innerPromise_'s resolution. It is performed by host environments as part of HostImportModuleRecordDynamically.</dd>
-            </dl>
-            <emu-alg>
-              1. Let _fulfilledClosure_ be a new Abstract Closure with parameters (_result_) that captures _moduleRecord_ and _promiseCapability_ and performs the following steps when called:
-                1. Assert: _result_ is *undefined*.
-                1. Assert: Evaluate has already been invoked on _moduleRecord_ and successfully completed.
-                1. Let _namespace_ be Completion(GetModuleNamespace(_moduleRecord_)).
-                1. If _namespace_ is an abrupt completion, then
-                  1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _namespace_.[[Value]] &raquo;).
-                1. Else,
-                  1. Perform ! Call(_promiseCapability_.[[Resolve]], *undefined*, &laquo; _namespace_.[[Value]] &raquo;).
-                1. Return ~unused~.
-              1. Let _onFulfilled_ be CreateBuiltinFunction(_fulfilledClosure_, 0, *""*, &laquo; &raquo;).
-              1. Let _rejectedClosure_ be a new Abstract Closure with parameters (_error_) that captures _promiseCapability_ and performs the following steps when called:
-                1. Perform ! Call(_promiseCapability_.[[Reject]], *undefined*, &laquo; _error_ &raquo;).
-                1. Return ~unused~.
-              1. Let _onRejected_ be CreateBuiltinFunction(_rejectedClosure_, 0, *""*, &laquo; &raquo;).
-              1. Perform PerformPromiseThen(_innerPromise_, _onFulfilled_, _onRejected_).
+          <emu-alg>
+            1. If _referrer_ is not a Source Text Module Record, _referrer_.[[ModuleInstance]] is *undefined*, or _referrer_.[[ModuleInstance]].[[ImportHook]] is *undefined*, then
+              1. Perform HostLoadImportedModule(_referrer_, _specifier_, _hostDefined_, _payload_).
               1. Return ~unused~.
-            </emu-alg>
-
-            <emu-note type="editor">
-                This Abstract Operation is analogous to FinishDynamicImport but for Module Records with an associated Module Instance.
-            </emu-note>
-          </emu-clause>
+            1. Let _importHookResult_ be Completion(Call(_referrer_.[[ModuleInstance]].[[ImportHook]], _referrer_.[[ModuleInstance]].[[HandlerValue]], « _specifier_ »).
```

Comment:
what if the given specifier is already pending on the referrer?

```js
source = new ModuleSource(`
    import('a')
    import('a')
`)
new Module(source, { importHook: a_hook_that_resolves_after_5_seconds })
```

will this hook be called twice?

Reply:
Done! Now the hooks are always called at most once, and parallel calls will wait for the same result.